servers/src/main/java/tachyon/master/next/IndexedSet.java

```diff
@@ -0,0 +1,225 @@
+/*
+ * Licensed to the University of California, Berkeley under one or more contributor license
+ * agreements. See the NOTICE file distributed with this work for additional information regarding
+ * copyright ownership. The ASF licenses this file to You under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package tachyon.master.next;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.HashMap;
+import java.util.Set;
+import java.util.HashSet;
+import java.lang.reflect.Field;
+
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Sets;
+
+/**
+ * A set of objects that can be indexed by specific fields of the object, and different IndexedSet
+ * instances can use different fields to index. The field type must be comparable. The field value
+ * must not be changed after being added to the set, otherwise, behavior for all operations is not
+ * specified.
+ *
+ * In operations that need field name as the parameter, if the object does not have the field, a
+ * RuntimeException will be thrown.
+ *
+ * This class is thread safe.
+ */
+public class IndexedSet<T> {
+  /** All objects in the set */
+  private final Set<T> mObjects;
+  /** Map from field index to an index of field related object in the internal lists */
+  private final Map<FieldIndex<T>, Integer> mIndexMap;
+  /** List of maps from fieldValue to set of T which is indexed by field name */
+  private final List<Map<Object, Set<T>>> mSetIndexedByFieldValue;
+  /** List of {@link Field} which is indexed by field name */
+  private final List<Field> mFields;
```

Comment:
This variable `mFileds` is no more used as a `List` but only as an synchronized lock.  Why not make it simply of type `Object` and rename it as `mLock` or something more specific?